### PR TITLE
Parametrize CollectRNASeqMetrics 3'/5' Bias Length

### PR DESCRIPTION
**Pull request by ramsey-watchmaker**
### Description

From #1625, parameterized the length in CollectRNASeqMetrics for calculating 3' and 5' bias.

- Added an option to specify an end bias into a transcript replacing the constant PRIME_BASES (100 nt) variable. The default of this parameter set to the previous constant value.
- Previously transcripts under 100 nt (constant) were filtered out.  The previous fixed end bias distance was 100 nt and was, as far as I could tell chosen based on the end bias.  Therefore I added the bias length to the filter length criteria.
- Added a unit test with a few cases to supporting this change.
----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [x] Added or modified tests to cover changes and any new functionality
- [x] Edited the README / documentation (if applicable)
- [x] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

**Review by kachulis — CHANGES_REQUESTED** [review]
Thanks for this @ramsey-watchmaker.  A few documentation requests and some refactoring of the tests, but generally looks good.

File: src/test/java/picard/analysis/CollectRnaSeqMetricsTest.java [hunk 1]
```diff
@@ -382,6 +382,109 @@ public void testTranscriptionStrandMetrics() throws Exception {
...
+        samWriter.close();
+
+        File metricsFile = File.createTempFile("tmp.", ".rna_metrics");
+        metricsFile.deleteOnExit();
+
+        // Generate the metrics.
+        RnaSeqMetrics metrics;
+        MetricsFile<RnaSeqMetrics, Comparable<?>> output;
+
+        // Case 1 - 25 nt in, ends should both be 0 coverage.
```
[the remark this answers is not in the document]
Comment (ramsey-watchmaker) [reply src/test/java/picard/analysis/CollectRnaSeqMetricsTest.java]: Added a DataProvider.  Thanks.

File: src/test/java/picard/analysis/CollectRnaSeqMetricsTest.java [hunk 2]
```diff
@@ -459,4 +563,31 @@ public File getRefFlatFile(String sequence) throws Exception {
 
         return refFlatFile;
     }
+
+    public File getSimpleFlatFile(String sequence) throws Exception {
```
[the remark this answers is not in the document]
Comment (ramsey-watchmaker) [reply src/test/java/picard/analysis/CollectRnaSeqMetricsTest.java]: Added a polymorphic method.

File: src/main/java/picard/analysis/CollectRnaSeqMetrics.java [hunk 3]
```diff
@@ -132,6 +132,9 @@
     @Argument(shortName="LEVEL", doc="The level(s) at which to accumulate metrics.  ")
     public Set<MetricAccumulationLevel> METRIC_ACCUMULATION_LEVEL = CollectionUtil.makeSet(MetricAccumulationLevel.ALL_READS);
 
+    @Argument(doc="The distance into a transcript over which 5' and 3' bias is calculated.")
+    public int END_BIAS_BASES = 100;
```
[the remark this answers is not in the document]
Comment (ramsey-watchmaker) [reply src/main/java/picard/analysis/CollectRnaSeqMetrics.java:136]: Added.  Please let me know if you think it is too wordy, or pretty to the point.

File: src/test/java/picard/analysis/CollectRnaSeqMetricsTest.java [hunk 4]
```diff
@@ -382,6 +382,109 @@ public void testTranscriptionStrandMetrics() throws Exception {
...
+
+        // Set seed so that strandedness is consistent among runs.
+        builder.setRandomSeed(0);
+        final int sequenceIndex = builder.getHeader().getSequenceIndex(sequence);
+
+        // Add 30 reads per transcript. Adds a shift into the start and ends of reads so we can assess
+        // the dynamic characteristic of the bias windows.
+
+        final int read_length = 150;
+        final Integer[] test_sizes = new Integer[] {
```
[the remark this answers is not in the document]
Comment (ramsey-watchmaker) [reply src/test/java/picard/analysis/CollectRnaSeqMetricsTest.java]: Added.

**Comment by ramsey-watchmaker** [comment]
Hi @kachulis .  Thanks for the comments.  I addressed them and individually responded to comments.  Tests on update seem to pass as well.

File: src/main/java/picard/analysis/RnaSeqMetrics.java [hunk 5]
```diff
@@ -121,21 +121,25 @@
     = CORRECT_STRAND_READS/(CORRECT_STRAND_READS + INCORRECT_STRAND_READS).  0 if library is not strand-specific. */
     public double PCT_CORRECT_STRAND_READS;
 
-    /** The median coefficient of variation (CV) or stdev/mean for coverage values of the 1000 most highly expressed transcripts. Ideal value = 0. */
+    /** The median coefficient of variation (CV) or stdev/mean for coverage values of the 1000 most highly expressed transcripts
+     * that have a length greater than the END_BIAS_BASES parameter. Ideal value = 0. */
```
Comment (kachulis) [inline src/main/java/picard/analysis/RnaSeqMetrics.java:125]: 👍

File: src/test/java/picard/analysis/CollectRnaSeqMetricsTest.java [hunk 6]
```diff
@@ -382,6 +382,90 @@ public void testTranscriptionStrandMetrics() throws Exception {
...
+                    0.684932
+                }
+        };
+    }
+
+    @Test(dataProvider = "endBiasValues")
+    public void testBiasEndBiasAdjust(int end_bias_size, double expected_3prime, double expected_5prime, double expected_ratio) throws Exception {
+        final String sequence = "chr1";
+
+        // Create some alignments that hit the ribosomal sequence, various parts of the gene, and intergenic.
```
Comment (kachulis) [inline src/test/java/picard/analysis/CollectRnaSeqMetricsTest.java:416]: This comment isn't really correct for this test (doesn't hit robosomal, or intergenic, for example).  Could you add a comment describing the setup you are using here instead?

File: src/test/java/picard/analysis/CollectRnaSeqMetricsTest.java [hunk 7]
```diff
@@ -382,6 +382,90 @@ public void testTranscriptionStrandMetrics() throws Exception {
...
+        builder.setRandomSeed(0);
+        final int sequenceIndex = builder.getHeader().getSequenceIndex(sequence);
+
+        final int read_length = 150;
+        final int n_fragments = 30;
+        final String read_cigar = read_length + "M";
+        final int five_prime_start = 51;
+        final int three_prime_start = 975 - read_length;
+        final String refFlatStart = "0";
+        final String refFlatEnd = "1000";
```
Comment (kachulis) [inline src/test/java/picard/analysis/CollectRnaSeqMetricsTest.java:429]: I think it could be a little clearer where these numbers are coming from with a small adjustment.

```suggestion
        final String txStart = 0;
        final String txEnd = 1000;
        final int five_prime_start = txStart + 51;
        final int three_prime_start = txEnd - 25 - read_length;
        final String refFlatStart = txStart.toString();
        final String refFlatEnd = tsEnd.toString();

```

File: src/test/java/picard/analysis/CollectRnaSeqMetricsTest.java [hunk 8]
```diff
@@ -459,4 +544,12 @@ public File getRefFlatFile(String sequence) throws Exception {
...
         return refFlatFile;
     }
+
+    public File getRefFlatFile(String sequence) throws Exception {
+        // Default refFlat file with a single gene containing two exons, one of which is overlapped by the
+        // ribosomal interval.
+
+        return getRefFlatFile(sequence, "49", "500", "74", "400", "2", "49,249", "200,500");
+    }
+
```
Comment (kachulis) [inline src/test/java/picard/analysis/CollectRnaSeqMetricsTest.java:554]: remove extra blank line please

**Review by kachulis — COMMENTED** [review]
@ramsey-watchmaker thanks for those changes.  Just a couple more minor nits and then should be good to go.

**Comment by ramsey-watchmaker** [comment]
Hi @kachulis.  Thanks for the comments.  All made sense and are integrated.

File: src/test/java/picard/analysis/CollectRnaSeqMetricsTest.java [hunk 9]
```diff
@@ -413,7 +413,7 @@ public void testTranscriptionStrandMetrics() throws Exception {
     public void testBiasEndBiasAdjust(int end_bias_size, double expected_3prime, double expected_5prime, double expected_ratio) throws Exception {
         final String sequence = "chr1";
 
-        // Create some alignments that hit the ribosomal sequence, various parts of the gene, and intergenic.
+        // Create alignments that are shifted by a fixed distance from the 3' and 5' end of a trnscript.
```
Comment (kachulis) [inline src/test/java/picard/analysis/CollectRnaSeqMetricsTest.java:416]: transcript

**Review by kachulis — APPROVED** [review]
looks good!  One small typo, then good to go once tests pass.  Thanks for this @ramsey-watchmaker!

**Review by kachulis — APPROVED** [review]

File: src/main/java/picard/analysis/CollectRnaSeqMetrics.java [hunk 10]
```diff
@@ -168,7 +171,7 @@ protected void setup(final SAMFileHeader header, final File samFile) {
 
         collector = new RnaSeqMetricsCollector(METRIC_ACCUMULATION_LEVEL, header.getReadGroups(), ribosomalBasesInitialValue,
                 geneOverlapDetector, ribosomalSequenceOverlapDetector, ignoredSequenceIndices, MINIMUM_LENGTH, STRAND_SPECIFICITY, RRNA_FRAGMENT_PERCENTAGE,
-                true);
+                true, END_BIAS_BASES);
```
Comment (pshapiro4broad) [inline src/main/java/picard/analysis/CollectRnaSeqMetrics.java:174]: This fixes the case inside picard itself, but if others are importing this code and depending on the original signature of the constructor, a backward compatible fix would be to overload the constructor and allow both forms; with and without the `END_BIAS_BASES` parameter.

Comment (ramsey-watchmaker) [reply src/main/java/picard/analysis/CollectRnaSeqMetrics.java:174]: Hi @pshapiro4broad .  That comment makes sense around backwards compatibility.

**Comment by gbggrant** [comment]
@ramsey-watchmaker can you update this branch from master?  When that's done I'll go ahead and merge it.

**Comment by kbergin** [comment]
Thank you very much for your contribution to Picard, @ramsey-watchmaker !
